fastrand.go

```diff
@@ -20,65 +20,62 @@ import (
 // is the concatenation of an initial seed and a 128-bit counter. Each time
 // the entropy is hashed, the counter is incremented.
 type randReader struct {
-	entropy chan []byte
-}
-
-// Read fills b with random data. It always returns len(b), nil.
-func (r *randReader) Read(b []byte) (int, error) {
-	n := 0
-	for n < len(b) {
-		n += copy(b[n:], <-r.entropy)
-	}
-	return n, nil
-}
-
-// fillEntropy continuously fills r.entropy with new entropy.
-func (r *randReader) fillEntropy() {
-	// Create a hasher and fill it with 64 bytes of entropy. Technically only 16
-	// should be needed, but the underlying RNG may not be secure.
-	h, _ := blake2b.New256(nil)
-	_, err := io.CopyN(h, rand.Reader, 64)
-	if err != nil {
-		panic("fastrand: no entropy available")
-	}
-	seed := h.Sum(nil)
-
-	for {
-		for i := uint64(0); i < math.MaxUint64; i++ {
-			// Update the seed.
-			*(*uint64)(unsafe.Pointer(&seed[0])) = i
-
-			// Hash the seed.
-			h.Reset()
-			h.Write(seed)
-
-			// Send the entropy down the entropy channel.
-			r.entropy <- h.Sum(nil)
-		}
-
-		// Re-seed the hasher. Use the entropy that existed previously,
-		// protecting against a compromised RNG.
-		h.Reset()
-		h.Write(seed[:])
-		seed = h.Sum(seed[:0])
-	}
+	progress uint64
+	entropy  [32]byte
 }
 
 // Reader is a global, shared instance of a cryptographically strong pseudo-
 // random generator. It uses blake2b as its hashing function. Reader is safe
 // for concurrent use by multiple goroutines.
 var Reader io.Reader
 
+// init provides the initial entropy for the reader that will seed all numbers
+// coming out of fastrand.
 func init() {
-	r := &randReader{
-		entropy: make(chan []byte, 1000),
-	}
-	for i := 0; i < runtime.NumCPU(); i++ {
-		go r.fillEntropy()
+	r := &randReader{}
+	n, err := rand.Read(r.entropy[:])
```

Comment:
following up on the other comment, I would make this `rand.Read(r.entropy[16:])` so that the counters are initialized to zero.